README.rst
.. image:: https://img.shields.io/github/tag/daq-tools/terkin-javascript.svg
    :target: https://github.com/daq-tools/terkin-javascript

|

.. figure:: https://ptrace.getkotori.org/2019-06-26_terkin-squirrel-small.png
    :target: https://ptrace.getkotori.org/2019-06-26_terkin-squirrel-small.png


######################################
Terkin Telemetry Client for JavaScript
######################################

*****
About
*****
Terkin Telemetry client library for JavaScript, written in TypeScript.

Terkin is a client-side library, framework and concept for doing telemetry on
embedded compute nodes usually resembling sensor nodes of a sensor network.


********
Synopsis
********
::

    var terkin = require('./terkin');

    var telemetry_node = new terkin.TelemetryNode(
        "https://daq.example.org/api",
        {
            "realm": "mqttkit-1",
            "network": "testdrive",
            "gateway": "terkin",
            "node": "js-node-01",
        }
    );

    var data = {"temperature": 42.84, "humidity": 83.01};
    telemetry_node.transmit(data);


***********
Development
***********


Setup
=====
::

    yarn install


Operation
=========
Compile TypeScript in ``./src`` folder to JavaScript in ``./build`` folder::

    yarn compile

Run example program to publish single measurement reading::

    yarn run-example

Run example program to publish synthesized sawtooth signal::

    yarn run-sawtooth

Hacking
=======
Continuously watch the filesystem and run example program on change::

    yarn watch-example

Continuously watch the filesystem and run sawtooth signal example program on change::

    yarn watch-sawtooth


----

********************
Content attributions
********************

The copyright of particular images and pictograms are held by their respective owners, unless otherwise noted.

Squirrel by Maria from the Noun Project. Creative Commons CC-BY.

- https://thenounproject.com/morika/
- https://thenounproject.com/term/squirrel/83208/


CHANGES.rst
*********
Changelog
*********


In progress
===========


2019-06-26 0.1.0
================
- Compiles and runs
- Make things work
- Add pure JavaScript example


src/terkin.ts
/**
 *
 * Terkin Telemetry Client
 * =======================
 *
 * About
 * -----
 * Terkin Telemetry client library for JavaScript, written in TypeScript.
 *
 * Terkin is a client-side library, framework and concept for doing telemetry on
 * embedded compute nodes usually resembling sensor nodes of a sensor network.
 *
 * https://github.com/daq-tools/terkin-javascript
 *
 */
import request from 'request';

export class TelemetryNode {
  /**
   *
   * Telemetry node client: Highlevel network participant API.
   *
   */

  api_uri: string;
  realm: string;
  network: string;
  gateway: string;
  node: string;

  channel_uri: string;

  client: TelemetryClient;

  constructor(api_uri, options) {
    this.api_uri = api_uri;
    this.realm   = options['realm'];
    this.network = options['network'];
    this.gateway = options['gateway'];
    this.node    = options['node'];

    this.channel_uri = `${this.api_uri}/${this.realm}/${this.network}/${this.gateway}/${this.node}`;
    console.log('Channel URI: ', this.channel_uri);

    this.client = new TelemetryClient(`${this.channel_uri}/data`);
  }

  transmit(data) {
    return this.client.transmit(data);
  }

}

export class TelemetryClient {
  /**
   *
   * Telemetry data client: Lowlevel API.
   *
   */
  
  uri: string;

  constructor(uri) {
    this.uri = uri;
  }

  transmit(data) {
    /**
     *
     * Submit telemetry data using HTTP POST request
     * Serialization: x-www-form-urlencoded
     *
     * https://github.com/request/request
     *
     */

    //$payload = http_build_query($data);

    // Use key 'http' even if you send the request to https://...
    const options = {
      method: 'POST',
      uri: this.uri,
      headers: {'Content-Type': 'application/x-www-form-urlencoded'},
      form: data,
    };

    const json_data = JSON.stringify(data);

    console.info(`Transmitting data ${json_data}.`);
    request(options, function(error, response, body) {
      if (error) {

        // Print the error if one occurred
        console.error(`Submitting telemetry data to '${options.uri}' failed: ${error}. data=${JSON.stringify(json_data)}`);

        // Print the response status code
        console.log('Response status:', response && response.statusCode);

        // Print the response.
        console.log('Response body:', body);

      } else {

        // Print the response.
        console.log('Request successful');
        console.log('Response body:', body);

      }
    });
    
  }

}


src/example.ts
// -*- coding: utf-8 -*-
/*
=======================================================
Terkin telemetry demo program for JavaScript/TypeScript
=======================================================

Documentation
-------------
https://github.com/daq-tools/terkin-javascript


Synopsis
--------
Run demonstration program from the command line::

    # Send fixed measurements "temperature" => 42.84, "humidity" => 83
    # to defined target sink.
    yarn run-demo

    # Send a periodic, slowly oscillating sawtooth signal.
    yarn run-sawtooth

*/

import process from 'process';

// Acquire HTTP API library.
import { TelemetryNode } from './terkin';


class Example {

  telemetry_node: TelemetryNode;

  constructor(telemetry_node) {
    this.telemetry_node = telemetry_node;
  }

  single() {
    // Emit single sample of an example payload.
    const data = {"temperature": 42.84, "humidity": 83.01};
    this.telemetry_node.transmit(data);

  }

  async sawtooth() {

    // Set timezone.
    //date_default_timezone_set("Europe/Berlin");

    // Emit sample of a sawtooth signal each second, periodically
    while (true) {
      const data = {second: new Date().getUTCSeconds()};
      this.telemetry_node.transmit(data);
      await this.sleep(1000);
    }

  }

  sleep(ms) {
    // https://stackoverflow.com/questions/14249506/how-can-i-wait-in-node-js-javascript-l-need-to-pause-for-a-period-of-time
    // https://github.com/nodejs/node-v0.x-archive/issues/7283
    return new Promise(resolve => {
      setTimeout(resolve, ms);
    })
  }

}

function main() {

  // Read command line arguments.
  let action;
  if (process.argv.length > 1) {
    action = process.argv[2];

  } else {
    action = 'single';
  }

  // Create TelemetryNode object handle.
  const telemetry:TelemetryNode = new TelemetryNode(
    "https://daq.example.org/api",
    {
      "realm": "mqttkit-1",
      "network": "testdrive",
      "gateway": "terkin",
      "node": "js-node-01",
    }
  );

  // Run example scenario.
  const example = new Example(telemetry);
  if (action == 'single') {
    example.single();

  } else if (action == 'sawtooth') {
    example.sawtooth();

  } else {
    console.error(`Action "${action}" undefined.`);

  }

}

main();
